### PR TITLE
Make reference counting more generic

### PR DESCRIPTION
**Pull request by da-woods**
This PR allows types to define their own `generate_incref` (etc) type functions, with `put_incref` (etc) deferring to those functions. The idea being that mixed C struct/Python types can be created more easily and make use of the existing code for generating reference counting.

Right now this PR defines the such functions for `PyObject` and `MemoryviewSlice`. It's really designed to support https://github.com/cython/cython/pull/3346, so should probably be reviewed before that PR.

This PR does shouldn't change anything user-visible (hopefully) - it's just a change in structure to make other changes easier.

### CODE REVIEW

File: Cython/Compiler/Code.py [hunk 1]
```diff
@@ -2060,131 +2062,94 @@ def entry_as_pyobject(self, entry):
...
-                    null_check and 'X' or '',
-                    entry.cname))
-            else:
-                self.putln("__Pyx_%sDECREF(%s); %s = 0;" % (
-                    null_check and 'X' or '',
-                    self.entry_as_pyobject(entry),
-                    entry.cname))
+    def get_var_nullcheck(self, entry):
+        # we usually want to put this into an is statement so it's a get,
+        # not a put
```
Comment (scoder) [inline Cython/Compiler/Code.py]: I don't think we use `get_…` elsewhere. It's just `error_goto_if()`, for example.

File: Cython/Compiler/PyrexTypes.py [hunk 2]
```diff
@@ -193,6 +193,8 @@ class PyrexType(BaseType):
     #  is_pythran_expr       boolean     Is Pythran expr
     #  is_numpy_buffer       boolean     Is Numpy array buffer
     #  has_attributes        boolean     Has C dot-selectable attributes
+    #  needs_xxxref          boolean     Needs code to be generated similar to incref/gotref/decref.
```
Comment (scoder) [inline Cython/Compiler/PyrexTypes.py]: `needs_refcounting`. `xxxref` is a very confusing and unclear name.

File: Cython/Compiler/Nodes.py [hunk 3]
```diff
@@ -6040,7 +6045,7 @@ def generate_execution_code(self, code):
             arg.free_temps(code)
         code.putln(
             code.error_goto_if_null(temp_result, self.pos))
-        code.put_gotref(temp_result)
+        code.put_gotref(temp_result, PyrexTypes.py_object_type)
```
Comment (scoder) [inline Cython/Compiler/Nodes.py]:
```suggestion
        code.put_gotref(temp_result, py_object_type)
```

File: Cython/Compiler/Nodes.py [hunk 4]
```diff
@@ -8729,11 +8734,7 @@ def cleanup_temps(self, code):
         if self.is_parallel and not self.is_nested_prange:
             code.putln("/* Clean up any temporaries */")
             for temp, type in sorted(self.temps):
-                if type.is_memoryviewslice:
-                    code.put_xdecref_memoryviewslice(temp, have_gil=False)
-                elif type.is_pyobject:
-                    code.put_xdecref(temp, type)
-                    code.putln("%s = NULL;" % temp)
+                code.put_xdecref_clear(temp, type, do_for_memoryviewslice=True, have_gil=False)
```
Comment (scoder) [inline Cython/Compiler/Nodes.py]: There is a slight difference here. `put_xdecref_clear()` generates code to first clear the variable and then decref it. I think that should often be ok, since the C compiler should generally be able to optimise away the strict ordering for local variables. But it's something that has been bothering me for a while since this difference isn't always explicit in the generator code. Not to be solved right now, but to keep in mind.

File: Cython/Compiler/PyrexTypes.py [hunk 5]
```diff
@@ -334,6 +337,29 @@ def _assign_from_py_code(self, source_code, result_code, error_pos, code,
...
+                                      self)
+        else:
+            return None
+
+    def _generate_xxxref_set_placeholder(self, cname, rhs_cname, *ignored_args, **ignored_kwds):
+        if self.needs_xxxref:
+            raise NotImplementedError("Ref-counting operation not yet implemented for type %s" %
+                                      self)
+        else:
+            return "%s = %s" % (cname, rhs_cname)
```
Comment (scoder) [inline Cython/Compiler/PyrexTypes.py]:
```suggestion
        return "%s = %s" % (cname, rhs_cname)
```

File: Cython/Compiler/PyrexTypes.py [hunk 6]
```diff
@@ -334,6 +337,29 @@ def _assign_from_py_code(self, source_code, result_code, error_pos, code,
             convert_call,
             code.error_goto_if(error_condition or self.error_condition(result_code), error_pos))
 
+    def _generate_xxxref_placeholder(self, *ignored_args, **ignored_kwds):
+        if self.needs_xxxref:
+            raise NotImplementedError("Ref-counting operation not yet implemented for type %s" %
+                                      self)
+        else:
+            return None
```
Comment (scoder) [inline Cython/Compiler/PyrexTypes.py]: Erm, no. I don't want `None` to accidentally appear in the generated source code anywhere. Maybe an empty string instead?
```suggestion
        return ''
```

File: Cython/Compiler/PyrexTypes.py [hunk 7]
```diff
@@ -334,6 +337,29 @@ def _assign_from_py_code(self, source_code, result_code, error_pos, code,
...
+        if self.needs_xxxref:
+            raise NotImplementedError("Ref-counting operation not yet implemented for type %s" %
+                                      self)
+        else:
+            return "%s = %s" % (cname, rhs_cname)
+
+    generate_incref = generate_xincref = generate_decref = generate_xdecref \
+        = generate_decref_clear = generate_xdecref_clear \
+        = generate_gotref = generate_xgotref = generate_giveref = generate_xgiveref \
+        = generate_nullcheck \
```
Comment (scoder) [inline Cython/Compiler/PyrexTypes.py]: This feels unrelated and doesn't seem to be doing the right thing. I would expect it to be used in a condition, and then, what is the correct fallback there? A condition needs _something_ to check.
I would just remove this method.

File: Cython/Compiler/PyrexTypes.py [hunk 8]
```diff
@@ -1039,6 +1069,34 @@ def specialize(self, values):
...
+    def generate_incref(self, cname, do_for_memoryviewslice=False, **kwds):
+        if do_for_memoryviewslice:
+            # FIXME use_utility_code
+            return "__PYX_INC_MEMVIEW(&%s, %d);" % (cname, int(kwds['have_gil']))
+
+    # decref however did look to always apply for memoryview slices
+    # with "have_gil" set to True by default
+    def generate_xdecref(self, cname, have_gil=True, **kwds):
+        # FIXME use_utility_code
+        return "__PYX_XDEC_MEMVIEW(&%s, %d);" % (cname, int(have_gil))
```
Comment (scoder) [inline Cython/Compiler/PyrexTypes.py]: Swallowing all `**kwargs` always has a code smell because it tends to hide bugs and typos.

File: Cython/Compiler/PyrexTypes.py [hunk 9]
```diff
@@ -1039,6 +1069,34 @@ def specialize(self, values):
...
+        return "%s %s.memview = NULL; %s.data = NULL;" % (
+                self.generate_xdecref(cname, do_for_memoryviewslice=True, **kwds),
+                cname, cname)
+
+    def generate_decref_clear(self, cname, **kwds):
+        # memoryviews don't currently distinguish between xdecref and decref
+        return self.generate_xdecref_clear(cname, **kwds)
+
+    # memoryviews don't participate in giveref/gotref
+    generate_gotref = generate_xgotref = generate_xgiveref = generate_giveref = lambda *args: None
```
Comment (scoder) [inline Cython/Compiler/PyrexTypes.py]: Again: `None` is dangerous here. Better use an empty string.
Also, not accepting `**kwargs` here is at least inconsistent. Why not use the existing dummy methods instead of a new lambda?

File: Cython/Compiler/PyrexTypes.py [hunk 10]
```diff
@@ -1189,6 +1248,74 @@ def global_init_code(self, entry, code):
...
+            if clear_before_decref:
+                if not nanny:
+                    X = ''  # CPython doesn't have a Py_XCLEAR()
+                return "%s_%sCLEAR(%s);" % (prefix, X, cname)
+            else:
+                return ("%s_%sDECREF(%s); %s = 0;" % (
+                    prefix, X, self.as_pyobject(cname), cname))
+        else:
+            return ("%s_%sDECREF(%s);" % (
+                prefix, X, self.as_pyobject(cname)))
```
Comment (scoder) [inline Cython/Compiler/PyrexTypes.py]: No need for parentheses around the whole expressions in both cases.

File: Cython/Compiler/Code.py [hunk 11]
```diff
@@ -2060,131 +2062,94 @@ def entry_as_pyobject(self, entry):
...
             return entry.cname
 
     def as_pyobject(self, cname, type):
         from .PyrexTypes import py_object_type, typecast
         return typecast(py_object_type, type, cname)
 
-    def put_gotref(self, cname):
-        self.putln("__Pyx_GOTREF(%s);" % cname)
+    def put_gotref(self, cname, type):
+        self.putln(type.generate_gotref(cname))
```
Comment (scoder) [inline Cython/Compiler/Code.py]: In all of these methods, the `putln()` call should be avoided if the type decides that there is nothing to write.

File: Cython/Compiler/Code.py [hunk 12]
```diff
@@ -2060,131 +2062,94 @@ def entry_as_pyobject(self, entry):
...
-        self.putln("__Pyx_XDECREF_SET(%s, %s);" % (cname, rhs_cname))
+    def put_var_xdecref_clear(self, entry, **kwds):
+        self._put_var_decref_clear(entry, null_check=True, **kwds)
 
-    def put_var_decref(self, entry):
-        if entry.type.is_pyobject:
-            self.putln("__Pyx_XDECREF(%s);" % self.entry_as_pyobject(entry))
+    def _put_var_decref_clear(self, entry, null_check, **kwds):
+        f = getattr(entry.type, "generate_%sdecref_clear" %
+                                (null_check and "x" or ""))
```
Comment (scoder) [inline Cython/Compiler/Code.py]: Please spell this out to allow code indexers to find the use of these two methods.

File: Cython/Compiler/Code.py (hunk 11 above)
Comment (da-woods) [reply Cython/Compiler/Code.py]: I did add an `if None` right at the start of `putln` to try to account for this because otherwise all these functions ended up with duplicated boilerplate (which we're definitely trying to avoid in other place).

Maybe it'd be better to leave `putln` as it was, but add `putln_if_not_none` and then use that instead?

File: Cython/Compiler/Nodes.py [hunk 13]
```diff
@@ -2072,26 +2072,33 @@ def generate_function_definitions(self, env, code):
...
+            if entry.type.is_pyobject:
+                #if not (not entry.is_arg or len(entry.cf_assignments) > 1):
+                if entry.is_arg and len(entry.cf_assignments) <= 1:
+                    continue
+            code.put_var_xdecref(entry,
+                                 do_for_memoryviewslice=True,
+                                 have_gil=not lenv.nogil)
+                        # TODO ideally should pick based on
+                        # entry.xdecref_cleanup but this doesn't seem to be set reliably
+                        # for regular variables
```
[the remark this answers is not in the document]
Comment (da-woods) [reply Cython/Compiler/Nodes.py]: https://github.com/cython/cython/blob/0d272a51338467a5918911bed068d96bd28ad425/Cython/Compiler/Code.py#L2176-L2179

The current ``put_var_decref` - which has been there for 9 years - has generated `xdecref` whether you called `put_var_decref` or `put_var_xdecref`. So this change doesn't actually degrade the generated code. If definitely crashes sometimes if you try to use `xdecref_cleanup`.

I'll can try to probe a bit deeper and see if there's something obvious missed that'll get the desired behaviour though. But if I succeed it'll be an improvement, rather than restoring the status quo.

File: Cython/Compiler/Nodes.py [hunk 14]
```diff
@@ -3409,7 +3415,7 @@ def generate_function_definitions(self, env, code):
         code.put_label(code.return_label)
         for entry in lenv.var_entries:
             if entry.is_arg and entry.type.is_pyobject:
-                code.put_var_decref(entry)
+                code.put_var_xdecref(entry)
```
[the remark this answers is not in the document]
Comment (da-woods) [reply Cython/Compiler/Nodes.py]: I think this is because `put_var_decref` historically always put `xdecref`. Again, I'll see if a bit more intelligence can be applied here, but this isn't a degradation.

File: Cython/Compiler/PyrexTypes.py (hunk 9 above)
Comment (da-woods) [reply Cython/Compiler/PyrexTypes.py]: The existing dummy methods are designed to fail for types with `needs_refcounting` (since if they end up called then they really should exist). I've switched it to an empty string though.

File: Cython/Compiler/PyrexTypes.py (hunk 7 above)
Comment (da-woods) [reply Cython/Compiler/PyrexTypes.py]: This was added because it was useful for the fastcall args change - the types could fail to generate correctly and there was a way of testing it. The fallback should be `1` (i.e. basic types can never fail to generate) but right now it isn't.

I think I'd like to keep something like this, but I'll see if I can make it clearer.

File: Cython/Compiler/Code.py (hunk 11 above)
Comment (scoder) [reply Cython/Compiler/Code.py]: There is a use case for `putln("")` to write a blank line, and I wouldn't rely on the result of `type.generate_*()` to always end up directly in `putln()`. So `putln()` (or a similar method) is the wrong place to handle this.

Why not pass the `code` writer on into the type method, and let that decide completely (by calling `putln()`) about how and what to generate? That has the advantage that the actual code generation is more encapsulated between code writer and type, so that it can be moved around between the two without caring about other usages elsewhere. Currently, anyone could call `type.generate_*()` and mishandle the result by accident.

**Comment by da-woods** [comment]
Right - I've done made the easy changes and I'll come back to the ones that need more thought.

**Comment by scoder** [comment]
Yes, I think that looks much better. A couple of keyword argument issues left according to travis, and I'll give it another review soon.

**Comment by da-woods** [comment]
Closed and opened just to rerun tests. I don't think the previous batch actually failed - I think one of the OS X/Clang ones just ran out of time.

File: Cython/Compiler/Code.py [hunk 15]
```diff
@@ -2060,131 +2060,97 @@ def entry_as_pyobject(self, entry):
         type = entry.type
         if (not entry.is_self_arg and not entry.type.is_complete()
             or entry.type.is_extension_type):
-            return "(PyObject *)" + entry.cname
+            return "(PyObject *)" + entry.cname # FIXME need to implement this
```
Comment (scoder) [inline Cython/Compiler/Code.py]: ?

File: Cython/Compiler/ExprNodes.py [hunk 16]
```diff
@@ -746,19 +746,20 @@ def release_temp_result(self, code):
...
     def make_owned_memoryviewslice(self, code):
         """
         Make sure we own the reference to this memoryview slice.
         """
+        # TODO ideally this would be shared with "make_owned_reference"
         if not self.result_in_temp():
-            code.put_incref_memoryviewslice(self.result(),
-                                            have_gil=self.in_nogil_context)
+            code.put_incref_memoryviewslice(self.result(), self.type,
+                                            have_gil=True)
```
Comment (scoder) [inline Cython/Compiler/ExprNodes.py]: Interesting. This looks like it was previously passing the wrong state to `have_gil`. But then, it's either unused (in which case we shouldn't need to pass `have_gil` at all), or it should still depend on `not self.in_nogil_context`, because this method shouldn't just assume we own the GIL.

File: Cython/Compiler/ExprNodes.py [hunk 17]
```diff
@@ -849,6 +845,53 @@ def free_subexpr_temps(self, code):
...
+    #    return type
+
+    def generate_decref_set(self, code, rhs):
+        code.put_decref_set(self.result(), self.ctype(), rhs)
+
+    def generate_xdecref_set(self, code, rhs):
+        code.put_xdecref_set(self.result(), self.ctype(), rhs)
+
+    def generate_guarded_decref_set(self, code, rhs):
+        # TODO don't like the name
```
Comment (scoder) [inline Cython/Compiler/ExprNodes.py]: What about making it a keyword argument to `generate_decref_set()`? E.g. `handle_null=True`?

File: Cython/Compiler/PyrexTypes.py [hunk 18]
```diff
@@ -334,6 +337,31 @@ def _assign_from_py_code(self, source_code, result_code, error_pos, code,
             convert_call,
             code.error_goto_if(error_condition or self.error_condition(result_code), error_pos))
 
+    def _generate_xxxref_placeholder(self, code, *ignored_args, **ignored_kwds):
```
Comment (scoder) [inline Cython/Compiler/PyrexTypes.py]: The same "xxx" naming issue that I previously mentioned applies here, too. I'd call this `_generate_dummy_refcounting` and `_generate_dummy_refcounting_assignment` for the setter.

File: Cython/Compiler/PyrexTypes.py [hunk 19]
```diff
@@ -1189,6 +1248,75 @@ def global_init_code(self, entry, code):
...
+            code.putln("Py_INCREF(%s);" % self.as_pyobject(cname))
+
+    def generate_xincref(self, code, cname, nanny):
+        if nanny:
+            code.putln("__Pyx_XINCREF(%s);" % self.as_pyobject(cname))
+        else:
+            code.putln("Py_XINCREF(%s);" % self.as_pyobject(cname))
+
+    def generate_decref(self, code, cname, nanny, have_gil):
+        # have_gil is for the benefit of memoryviewslice - it's ignored here
```
Comment (scoder) [inline Cython/Compiler/PyrexTypes.py]: Why not also `assert have_gil`, just in case? Might detect a bug or two every once in a while. (Assuming it is always properly passed – if not, then that's obviously a bad idea.)

File: Cython/Compiler/Code.py (hunk 15 above)
Comment (da-woods) [reply Cython/Compiler/Code.py]: Apart from one call in `ModuleNode.py` this function's now basically unused. Instead I use `PyrexTypes.PyObjectType.as_pyobject()`. However, that's missing the the `not entry.is_self_arg` because I couldn't find a good way of passing it in. The comment was a reminder to look at that again.

Since it's a fairly minor bit of generated code-tidiness (it only avoids an extra cast, and I couldn't work out exactly what causes `is_self_arg` to be set anyway) I think it's simplest just to leave it out (and delete the comment).

File: Cython/Compiler/ExprNodes.py (hunk 16 above)
Comment (da-woods) [reply Cython/Compiler/ExprNodes.py]: I think this is a mistake and should be `have_gil=self.in_nogil_context`. Nothing seems to be failing with it changed back (but I'll see when it goes through the whole test-suite).

File: Cython/Compiler/ExprNodes.py [hunk 20]
```diff
@@ -2472,21 +2498,11 @@ def generate_deletion_code(self, code, ignore_nonexisting=False):
...
-                    else:
-                        code.put_decref(self.result(), self.ctype())
-                    code.putln('%s = NULL;' % self.result())
-                else:
-                    code.put_xdecref_memoryviewslice(self.entry.cname,
-                                                     have_gil=not self.nogil)
+                if self.entry.in_closure:
+                    # generator
+                    self.generate_gotref(code, handle_null=True, maybe_null_extra_check=ignore_nonexisting)
+                code.put_xdecref_clear(self.result(), self.ctype(),
```
Comment (scoder) [inline Cython/Compiler/ExprNodes.py]: There used to be a decref (without x) here if `ignore_nonexisting and self.cf_maybe_null`. So this is a regression.

I don't care so much about giveref/gotref with or without x. They are just debug code that vanishes in production builds. I do care for incref/decref, though.

File: Cython/Compiler/Nodes.py [hunk 21]
```diff
@@ -2072,26 +2071,28 @@ def generate_function_definitions(self, env, code):
...
-                        code.put_var_xdecref(entry)
-                    else:
-                        code.put_var_decref(entry)
+            if entry.type.is_pyobject:
+                if entry.is_arg and not entry.cf_is_reassigned:
+                    continue
+            if entry.xdecref_cleanup:
+                code.put_var_xdecref(entry, have_gil=not lenv.nogil)
+            else:
+                code.put_var_xdecref(entry, have_gil=not lenv.nogil)
```
Comment (scoder) [inline Cython/Compiler/Nodes.py]: This code screams for either a clear warning comment, or an actual `put_var_decref()` in the second case.

File: Cython/Compiler/Nodes.py [hunk 22]
```diff
@@ -2072,26 +2071,28 @@ def generate_function_definitions(self, env, code):
...
-                                                 have_gil=not lenv.nogil)
+                if is_cdef and not entry.cf_is_reassigned:
+                    continue
+            else:
+                if entry.in_closure:
+                    continue
+                if not acquire_gil and not entry.cf_is_reassigned:
+                    continue
+
+            code.put_var_xdecref(entry, have_gil=not lenv.nogil)
```
Comment (scoder) [inline Cython/Compiler/Nodes.py]:
```suggestion
            # FIXME: try to use `put_var_decref()` here instead.
            code.put_var_xdecref(entry, have_gil=not lenv.nogil)
```

File: Cython/Compiler/ExprNodes.py [hunk 23]
```diff
@@ -849,6 +845,47 @@ def free_subexpr_temps(self, code):
...
+    #    type = self.type
+    #    if self.is_temp and self.type.is_pyobject:
+            # when a temp object is created it's always just
+            # created as a py_object_type and not a
+            # derived type
+    #        type = PyrexTypes.py_object_type
+    #    return type
+
+    def generate_decref_set(self, code, rhs, handle_null=False):
+        if not (handle_null and self.cf_is_null):
```
[the remark this answers is not in the document]
Comment (da-woods) [reply Cython/Compiler/ExprNodes.py]: This replicates the previous behaviour in the place it's used https://github.com/cython/cython/blob/4dd7c18de76cb5c90a5168fba35990600307e642/Cython/Compiler/ExprNodes.py#L2358-L2367.

However, since it's only used once and the behaviour doesn't make sense for `generate_decref_set` I think I'll remove the `handle_null` option from `generate_decref_set` and just return the place where it's used to doing it explicitly.

File: Cython/Compiler/Nodes.py (hunk 21 above)
Comment (da-woods) [reply Cython/Compiler/Nodes.py]: Unfortunately it seems to need to be like this. I've created an issue https://github.com/cython/cython/issues/3455 and put a warning comment in for now

File: Cython/Compiler/PyrexTypes.py [hunk 24]
```diff
@@ -1085,6 +1085,9 @@ def generate_incref_memoryviewslice(self, code, slice_cname, have_gil):
     def generate_xdecref(self, code, cname, nanny, have_gil):
         code.putln("__PYX_XDEC_MEMVIEW(&%s, %d);" % (cname, int(have_gil)))
 
+    def generate_decref(self, code, cname, nanny, have_gil):
+        self.generate_xdecref(code, cname, nanny, have_gil)
```
Comment (scoder) [inline Cython/Compiler/PyrexTypes.py]: This needs a comment, too. Otherwise, someone will see it and spend a lot of time trying to fix the "obvious typo", which is probably not just a typo.

File: Cython/Compiler/ExprNodes.py [hunk 25]
```diff
@@ -759,7 +759,7 @@ def make_owned_memoryviewslice(self, code):
         # TODO ideally this would be shared with "make_owned_reference"
         if not self.result_in_temp():
             code.put_incref_memoryviewslice(self.result(), self.type,
-                                            have_gil=True)
+                                            have_gil=self.in_nogil_context)
```
Comment (scoder) [inline Cython/Compiler/ExprNodes.py]: It apparently makes no difference (i.e. probably isn't used in the tests), but I think it needs to be negated.
```suggestion
                                            have_gil=not self.in_nogil_context)
```

**Comment by scoder** [comment]
Thanks a lot @da-woods !
